### PR TITLE
feat(table): add component package

### PR DESCRIPTION
**Pull request by richbachman**
- [x] Added Table component, stories, and tests
- [x] Added necessary props to Box
- [x] Added word-wrap to Text
- [x] Added Table to component Core

### CODE REVIEW

File: packages/paste-core/components/table/src/Tr.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    <Box
+      {...safelySpreadBoxProps(props)}
+      ref={ref}
+      as="tr"
+      borderStyle="solid"
+      borderColor="colorBorderLighter"
+      borderWidth="borderWidth0"
+      borderBottomWidth="borderWidth10"
+      verticalAlign={verticalAlignment}
+      _even={{backgroundColor: 'colorBackgroundRowStriped'}}
```
Comment (serifluous) [inline packages/paste-core/components/table/src/Tr.tsx:16]: We could potentially deprecate this token, and use colorBackground here instead.

@SiTaggart originally added this token, so will wait for him to chime in.

Comment (SiTaggart) [reply packages/paste-core/components/table/src/Tr.tsx:16]: It depends on if you ever feel like you'd like specific control over row striping (across all components, not just tables), independent from a generic background.

Would a dark theme potentially need it? Doubt it, but you never know.

File: packages/paste-core/components/table/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import * as PropTypes from 'prop-types';
+import {WidthOptions, isWidthTokenProp} from '@twilio-paste/style-props';
+
+export type AlignmentOptions = 'left' | 'center' | 'right';
+export type TableVariants = 'auto' | 'fixed';
+export type VerticalAlignmentOptions = 'top' | 'middle' | 'bottom';
```
Comment (SiTaggart) [inline packages/paste-core/components/table/src/types.ts:6]: You could potentially keep these types in sync with the prop-types below by using constants.

Then you can derive the types from the constant, and populate the oneOf values from the constant too. That way it's a single update in the future to keep them in sync with each other

File: packages/paste-style-props/src/types/typography.ts [hunk 3]
```diff
@@ -55,4 +57,5 @@ export interface TypographyProps {
   textTransform?: TextTransform;
   whiteSpace?: WhiteSpace;
   wordBreak?: WordBreak;
+  wordWrap?: WordWrap;
```
Comment (SiTaggart) [inline packages/paste-style-props/src/types/typography.ts:60]: Is there a prop-type that also needs this? I can't remember if we do prop types for typography

File: packages/paste-core/components/table/src/Td.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      fontSize="fontSize30"
+      lineHeight="lineHeight30"
+      paddingBottom="space40"
+      paddingLeft="space40"
+      paddingRight="space40"
+      paddingTop="space40"
+      position="relative"
+      textAlign={alignment}
+      verticalAlign="inherit"
+      width={width}
```
Comment (SiTaggart) [inline packages/paste-core/components/table/src/Td.tsx:20]: I'm tempted to only have width set on the column header. Setting width on a cell seems odd, and limiting to headers focuses the size setting API to one area of a table. Thoughts?

Comment (richbachman) [reply packages/paste-core/components/table/src/Td.tsx:20]: Yeah. I think thats a good idea. Seems weird to set the width in multiple areas.

File: packages/paste-style-props/src/types/typography.ts (hunk 3 above)
Comment (richbachman) [reply packages/paste-style-props/src/types/typography.ts:60]: Yep, but forgot to add it to the typography prop types, so good catch. I just added it.

File: packages/paste-core/components/table/src/Table.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import * as React from 'react';
+import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
+import {TableProps, TablePropTypes} from './types';
+
+const Table = React.forwardRef<HTMLTableElement, TableProps>(({variant = 'auto', ...props}, ref) => {
```
Comment (TheSisb) [inline packages/paste-core/components/table/src/Table.tsx:5]: Just wondering if, for simplicity, we should call the `variant` prop `tableLayout`.  Similarly for the other parts of the table: `alignment` -> `textAlign`, `verticalAlignment` -> `verticalAlign`.

Reason being that it's a familiar API for folks who know how HTML tables work and we aren't really doing anything special with these props.  What do you think?

Comment (richbachman) [reply packages/paste-core/components/table/src/Table.tsx:5]: I like that better. More familiar API, and it also opens up `variant` if we ever need that in the future for say an inverse table.

File: packages/paste-core/components/table/src/types.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+export interface TrProps {
+  children: NonNullable<React.ReactNode>;
+  verticalAlign?: TableVerticalAlignmentOptions;
+}
+
+export const TrPropTypes = {
+  children: PropTypes.node.isRequired,
+  verticalAlign: PropTypes.oneOf(
+    Object.keys(TableVerticalAlignmentObject).map(verticalAlignment => TableVerticalAlignmentObject[verticalAlignment])
```
Comment (SiTaggart) [inline packages/paste-core/components/table/src/types.ts:44]: can't you do `Object.values(TableVerticalAlignmentObject)` for these?

File: packages/paste-core/components/table/stories/index.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+import {MoreIcon} from '@twilio-paste/icons/esm/MoreIcon';
+import {Text} from '@twilio-paste/text';
+import {Truncate} from '@twilio-paste/truncate';
+import {Table, THead, TBody, TFoot, Tr, Td, Th} from '../src';
+
+const MoreMenu: React.FC<{}> = () => {
+  const menu = useMenuState();
+  return (
+    <Box display="flex" justifyContent="flex-end">
+      <MenuButton {...menu} tableLayout="reset" size="reset">
```
Comment (SiTaggart) [inline packages/paste-core/components/table/stories/index.stories.tsx:17]: Not sure this is supposed to be `tableLayout`. It's given you some funky blue icon buttons

image

Comment (richbachman) [reply packages/paste-core/components/table/stories/index.stories.tsx:17]: Damn find/replace all!

File: packages/paste-core/components/table/src/types.ts (hunk 6 above)
Comment (richbachman) [reply packages/paste-core/components/table/src/types.ts:44]: Much simpler.

**Review by SiTaggart — APPROVED** [review]

File: packages/paste-core/components/table/stories/index.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+              </Text>
+            </Td>
+            <Td>Incoming</Td>
+            <Td>(602) 609-6747</Td>
+          </Tr>
+        </TBody>
+      </Table>
+    );
+  })
+  .add('Fixed table', () => {
```
Comment (TheSisb) [inline packages/paste-core/components/table/stories/index.stories.tsx:96]: The layperson may read this as "not broken". We should be a little more specific that this is a `tableLayout: fixed` Table

File: packages/paste-core/components/table/stories/index.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+            <Td>&nbsp;</Td>
+            <Td>Cell</Td>
+            <Td textAlign="center">Content</Td>
+            <Td textAlign="right">3</Td>
+          </Tr>
+        </TFoot>
+      </Table>
+    );
+  })
+  .add('Row textAlign', () => {
```
Comment (TheSisb) [inline packages/paste-core/components/table/stories/index.stories.tsx:255]: Isn't this example still a column textAlign? 😕

File: packages/paste-core/components/table/stories/index.stories.tsx (hunk 8 above)
Comment (richbachman) [reply packages/paste-core/components/table/stories/index.stories.tsx:96]: Changing to `TableLayout: Fixed table`.

File: packages/paste-core/components/table/stories/index.stories.tsx (hunk 9 above)
Comment (richbachman) [reply packages/paste-core/components/table/stories/index.stories.tsx:255]: I changed this one to `Row verticalAlign` to be more clear.

**Review by TheSisb — APPROVED** [review]

**Pull request merged**